src/wh_server.c

```diff
@@ -28,8 +30,613 @@
+    if (ret > 0 ) {
+        meta->id = ret;
+        ret = hsmCacheKey(ctx, meta, keyBuf);
+    }
+    return ret;
+}
+
+static int hsmLoadKeyRsa(WOLFHSM_CTX* ctx, NvmMetaData* meta, RsaKey* key)
+{
+    int ret;
+    uint32_t idx = 0;
+    byte keyBuf[WOLFHSM_KEYSIZE];
+    /* retrieve the key */
+    meta->len = WOLFHSM_KEYSIZE;
+    ret = hsmReadKey(ctx, meta, keyBuf);
+    /* decode the key */
+    if (ret == 0)
+        ret = wc_RsaPrivateKeyDecode(keyBuf, &idx, key, meta->len);
+
+    return ret;
+}
+
+static int hsmCacheKeyEcc(WOLFHSM_CTX* ctx, NvmMetaData* meta, ecc_key* key)
+{
+    int ret;
+    uint32_t qxLen;
+    uint32_t qyLen;
+    uint32_t qdLen;
+    byte keyBuf[WOLFHSM_KEYSIZE];
+    /* export key */
+    qxLen = qyLen = qdLen = key->dp->size;
+    ret = wc_ecc_export_private_raw(key, keyBuf, &qxLen,
+        keyBuf + qxLen, &qyLen, keyBuf + qxLen + qyLen, &qdLen);
+    /* cache key */
+    if (ret == 0) {
+        ret = hsmGetUniqueId(ctx);
+    }
+    if (ret > 0) {
+        meta->len = qxLen + qyLen + qdLen;
+        meta->id = ret;
+        ret = hsmCacheKey(ctx, meta, keyBuf);
+    }
+    return ret;
+}
+
+static int hsmLoadKeyEcc(WOLFHSM_CTX* ctx, NvmMetaData* meta, ecc_key* key,
+    int curveId)
+{
+    int ret;
+    uint32_t keySz;
+    byte keyBuf[WOLFHSM_KEYSIZE];
+    /* retrieve the key */
+    meta->len = WOLFHSM_KEYSIZE;
+    ret = hsmReadKey(ctx, meta, keyBuf);
+    /* decode the key */
+    if (ret == 0) {
+        keySz = meta->len / 3;
+        ret = wc_ecc_import_unsigned(key, keyBuf, keyBuf + keySz,
+            keyBuf + keySz * 2, curveId);
+    }
+    return ret;
+}
+#endif
+
+static int hsmGetUniqueId(whServerContext* server)
+{
+    int i;
+    int ret = 0;
+    uint16_t id;
+    /* try every index, unless both cache and nvm are full one will work */
+    for (id = 1; id < WOLFHSM_NUM_RAMKEYS + 1; id++) {
+        /* check against chache keys */
+        for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+            if (id == server->cache[i].meta->id)
+                break;
+        }
+        /* try again if match */
+        if (i < WOLFHSM_CACHE_COUNT)
+            continue;
+#if 0
+        /* check against nvm keys */
+        for (i = 0; i < WOLFHSM_KEYSLOT_COUNT && ret == 0; i++) {
+            if (ctx->nvmMetaCache[i].id == id) {
+                break;
+            }
+        }
+        /* try again if match */
+        if (i < WOLFHSM_KEYSLOT_COUNT)
+            continue;
+#endif
+        /* break if our id didn't match either cache or nvm */
+        break;
+    }
+    /* if we've run out of space, return MEMORY_E */
+    if (ret == 0 && id >= WOLFHSM_NUM_RAMKEYS + 1)
+        ret = MEMORY_E;
+    /* ultimately, return found id */
+    if (ret == 0)
+        ret = id;
+    return ret;
+}
+
+static int hsmCacheKey(whServerContext* server, NvmMetaData* meta, uint8_t* in)
+{
+    int i;
+    int foundIndex = -1;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
```

Comment:
This comment is a catch-all for the whole file: I think we should be using wolfHSM error codes, not wolfSSL error codes. I think we only want to use wolfCrypt error codes to process result of wolfCrypt  functions

Reply:
wolfCrypt errors should be returned from cryptocb.c, but wolfHSM everywhere else.  Our goal was to have very few error codes to make autonomous handling simple.  Broke or not broke.

Reply:
Added to todo list